### ISSUE TITLE
Create proposal for Intermittent Welds

### ISSUE THREAD

**Issue by DrCaFr** [economidis-nick/createXSDforxMCF#53]
#### Cited from @mweinert611's slide set "03_WorkingMaterial_xMCF_WG_Meeting_201210.pdf":  
![grafik](https://user-images.githubusercontent.com/18173895/106766274-02709100-663a-11eb-8042-06c4807e4d8c.png)

---

#### Cited from @DrCaFr's personal minutes:  
![grafik](https://user-images.githubusercontent.com/18173895/106767003-b8d47600-663a-11eb-9716-f94425aee5c0.png)  
![grafik](https://user-images.githubusercontent.com/18173895/106767094-d1dd2700-663a-11eb-8242-d86a5884efd2.png)  
![grafik](https://user-images.githubusercontent.com/18173895/106767288-fc2ee480-663a-11eb-8808-bfed61664ed4.png)

---

#### Cited from @mweinert611's protocol (2020-12-17):  

![grafik](https://user-images.githubusercontent.com/18173895/106765418-123ba580-6639-11eb-8253-66f2614a47d3.png)

Small group around Nick Economidis (@economidis-nick) and Carsten Franke (@DrCaFr) shall prepare proposals for the open items for
- Step welds

**Comment by DrCaFr** [comment]
Let this issue be our next topic: 

@economidis-nick , @DrCaFr, @TimGuirguis, @mweinert611, Kosmas, @StephanHaasECS 

Karin to be asked for participation.

Next meeting: 2021-02-05, 10-11 CET. @DrCaFr will send invitation.

Home work: Bring some ideas ...

**Comment by DrCaFr** [comment]
Current definition of `<loc_list/>` is:  
![grafik](https://user-images.githubusercontent.com/18173895/106740978-6768be80-661b-11eb-95b0-453f672c8d34.png)

The topic/phrase "A stepped connection line" is addressed at this occurrence, only.

**Comment by DrCaFr** [comment]
Quotation from the protocol of 2015-11-26 meeting:
> 
-	Descriptions for Step welds 
	-	All descriptions will be removed for V3.0 but needs to be re-introduced for later with better description
	-	Responsible: J. Golumba
>

**Comment by DrCaFr** [comment]
Quotation from 2015-10-23 draft of &chi;MCF 3.0 document:  
> 
![grafik](https://user-images.githubusercontent.com/18173895/106743793-1c50aa80-661f-11eb-8d6d-a999e29ab7d9.png)
> 
![grafik](https://user-images.githubusercontent.com/18173895/106743848-2f637a80-661f-11eb-90d1-f3b02acbcbba.png)

**Comment by DrCaFr** [comment]
During 2015 discussions, Peter Mikolaj's document "FAT_AK25_MCF_Seamweld_Extensions_V421.doc" from 2010-05-03 was referenced. It contained paragraphs like these: 

![grafik](https://user-images.githubusercontent.com/18173895/106747401-27f2a000-6624-11eb-92b7-6040ca90eb58.png)

![grafik](https://user-images.githubusercontent.com/18173895/106747647-7c961b00-6624-11eb-9c32-ffa05f796a85.png)

![grafik](https://user-images.githubusercontent.com/18173895/106747822-b109d700-6624-11eb-9778-070db1bd85de.png)

**Comment by DrCaFr** [comment]
Jürgen points out that "Step Weld" denotes a specific sequence of spotwelds.  
The German notion "unterbrochene Schweißnaht" would require another translation.   
https://dict.leo.org suggests "intermittent weld".

**Comment by DrCaFr** [comment]
#### Comment by Karin:

From the fatigue analyst's point of view following remarks should be taken into account:

* in the major number of cases welding start and welding end are   the most important areas of the seamweld

* begin and also the end of the welding should be defined precisely  i.e. by a certain and fixed margin followed by a complete welding step

* the amount of how much of the welding line is covered by welded  sections could be expressed by some kind of "density"

* of course, the number of steps is also an important parameter

Following these considerations we can define the situation with 3 parameters:
- the margin (m)
- the number of steps (n)
- the stepping density ( d = l/(l+s) )

with the length (l) of the welding step and (s) the space between the steps

(l) and (s) then can be determined by the relation

L - 2m = n(l+s) -s  

where L is the overall length given by the length of the curve resulting from `<loc_list/>`.

From this relation we get 

l = d K,       s = (d-1)K     with    K = (L-2m)/(n+d-1)

With the 3 parameters mentioned above it should be possible to uniquely define the step weld.
![grafik](https://user-images.githubusercontent.com/18173895/107003992-e8e05e00-678d-11eb-8f13-cb470db746e0.png)

**Comment by DrCaFr** [comment]
#### Agenda for 2021-02-05 meeting:

1. Naming: "intermittent weld" instead of "step weld".
2. Review of considerations of previous years.
3. (automated) adaption of step parameters to avoid chopping of tiny pieces at the end.
4. Misc.
   - Next meeting(s) - _not_ on Fri. to allow Jürgen to join

**Comment by DrCaFr** [comment]
For the naming, candidates are:
- stitch weld
- skip weld
- intermittent weld

Contra "stitch weld": This would imply _regularity_.

Pro "intermittent weld": See https://en.wikipedia.org/wiki/Fillet_weld

__Decision:__ Let's use "intermittent weld".

**Comment by DrCaFr** [comment]
We need both (mututally exclusive) variants of intermittenmd wleds: reguklar _and_ irregular.

**Comment by DrCaFr** [comment]
We have to acknowledge to requirements wrt. avoiding chopping of tiny pieces at the end:
- keep the length, adapt the sequence
- keep the sequence, adapt the length

**Comment by DrCaFr** [comment]
Benefit of intermittent welds compared with (lots of) individual weld lines is _reduction of administrative overhead_.

**Comment by DrCaFr** [comment]
__Home work:__ Prepare some written proposal for all three options: Irregular, regular.a & regular.b. ==> @economidis-nick (Only for regular, since for irregular, we can fall back to the 2015-10-23 approach).

__Next meeting:__ Thursday, Feb. 18th, 10-11.

**Comment by economidis-nick** [comment]
### proposal for option 'regular.a':
In this proposal, user requires that stitch weld is described by the following parameters:
* number of steps along the weld (`n`)
* density (`d`) of steps, where `d = l / (l+s)`
* margins (`m1`, `m2`) from begining and end of seam

![image](https://user-images.githubusercontent.com/33229181/107058106-da6d6300-67dc-11eb-81ad-ee74150e35c2.png)
![image](https://user-images.githubusercontent.com/33229181/107058564-713a1f80-67dd-11eb-8fe0-ec8c28136148.png)



### proposal for option 'regular.b':
In this proposal, user requires that stitch weld is described by the following parameters:
* length(`l`) of each step
* spacing(`s`) between each step
* margins (`m1`, `m2`) from begining and end of seam

With a given curve (of length Ltotal) and an integral number (n) of steps in the stitch, there is a rounding 'error' that must be absorbed by step length and/or spacing.

Therefore, we must provide 3 options:
* `l` absorbs the rounding, modifying it to `l'`
* '`s`' absorbs the rounding, modifying it to `s'`
* both '`l`' and '`s`' absorb the rounding, keeping their relative ratio _constant_.
 Therefore, density `d` can be calculated.
  This makes the problem the same as **regular.a** !


![image](https://user-images.githubusercontent.com/33229181/107059842-f3771380-67de-11eb-849b-83c31f6732c7.png)



This leads to a proposal for xml:
#### regular.a
~~~~xml
<seamweld>
    <corner_weld base="1" technology="arc">
        <weld_position x='0.0' y='10.0' z='10.0' reference = 'no' angle='90' filler='yes' shape='convex'>
                <regular_step begin_margin="1.0" end_margin="3.0" density="0.666" num_steps="30"/>
        </weld_position>
    </corner_weld>
</seamweld>
~~~~

#### regular.b
~~~~xml
<seamweld>
    <corner_weld base="1" technology="arc">
        <weld_position x='0.0' y='10.0' z='10.0' reference = 'no' angle='90' filler='yes' shape='convex'>
                <regular_step begin_margin="1.0" end_margin="3.0" length="10.0" spacing="5.0"/>
        </weld_position>
    </corner_weld>
</seamweld>
~~~~

#### irregular intermittent welds
~~~~xml
<seamweld>
    <corner_weld base="1" technology="arc">
        <weld_position x='0.0' y='10.0' z='10.0' reference = 'no' angle='90' filler='yes' shape='convex'>
             <step_list>
                 <step from=30" to="140" />
                 <step from=210" to="350" />
                 <step from=410" to="550" />
             </step_list>
        </weld_position>
    </corner_weld>
</seamweld>
~~~~

**Comment by mweinert611** [comment]
What about using a parameter "**adjust**" in both cases (a,b) with the options
> case a:  "begin_margin","end_margin","densitiy"  (default="density")
> case b:  "begin_margin", "end_margin", "length", "spacing","density"; (default="spacing")

**Comment by economidis-nick** [comment]
> What about using a parameter "adjust" in both cases (a,b) with the options

* For **case a**, there is no need for explicit adjustment. `length` and `spacing` are always derived from `num_steps` and `density`. Neither `density`, nor `num_steps` ever are ajusted, or need adjusting.

* For **case b**, you are right, I forgot to add the adjustment parameter.
Note that my proposal only works by adjusting `spacing`, `length`, or `spacing_and_length`.
`margin1` not `margin2` could **not** be adjusted to absorb the excess length of `N*(s+l)-l` (where `N` is integer):
This excess could be greater that the margin => there is still an excess, but margin may not be able to absorb it.

I think I will stick to providing  _only_ these 3 options for _case b_
* `adjust = 'spacing'`
* `adjust = 'length'` and
* `adjust = 'length_and_spacing'`

### regular.b
~~~~xml
<seamweld>
    <corner_weld base="1" technology="arc">
        <weld_position x='0.0' y='10.0' z='10.0' reference = 'no' angle='90' filler='yes' shape='convex'>
                <regular_step begin_margin="1.0" end_margin="3.0" length="10.0" spacing="5.0" adjust="spacing"/>
        </weld_position>
    </corner_weld>
</seamweld>
~~~~

**Comment by StephanHaasECS** [comment]
- For **case a** I agree with Nick, if the total length somehow changes _**length**_ and _**spacing**_ are automatically adjusted. There is no need for an additional parameter
- For **case b** I think the chain of thought would be "What do I want to primarily keep of the parameters". So my suggestion would be to use something like **_keep_** instead of **_adjust_**. "keep = spacing", "keep = length" and "keep = density" (**_density_** being the value that is kept constant if you adjust both **_length_** and **_spacing_**).
Possible defaultvalues would be either **_spacing_** or **_density_** in my opinion. With an attribute called **_keep_** I would tend towards **_density_** as default. The reasoning being that if the user does not give a parameter to keep I would try to keep the adjustments as small as possible by adjusting **_spacing_** and **_length_**.
It would even be possible to reduce the possible values to "keep = spacing" and "keep = length" and if no value is given to then adjust  both **_spacing_** and **_length_** by default.
If we stick with **_adjust_**, **_spacing_** sounds good as default

I also think we need to specify how we want to describe the use cases later. Our summary from last time is:
> We have to acknowledge requirements wrt. avoiding chopping of tiny pieces at the end:
> a. keep the total length, adapt the sequence
> b. keep the sequence, adapt the total length

In my opinion both use cases for regular welded seams keep the total length and adapt the sequence in different ways. The total length of the intermittent weld is given by the **_<loc_list/>_** and the best possible interpretation the programs are able to do of it (e.g. in a given mesh)
- **case a** focuses on a given number of weld steps
- **case b** focuses more on getting the individual steps as close as possible to the demand

**Comment by economidis-nick** [comment]
> It would even be possible to reduce the possible values to "`keep = spacing`" and "`keep = length`" and if no value is given to then adjust both spacing and length by default.
I like Stephan's proposal. It emphasizes the intention (keep spacing/length/density).
We should specify it carefully though, otherwise one might get the impression that `margin` absorbs the rounding !

**Comment by DrCaFr** [comment]
> We have to acknowledge requirements wrt. avoiding chopping of tiny pieces at the end.
> Apparently, they can be addressed in two ways:
> a. keep the total length, adapt the sequence
> b. keep the sequence, adapt the total length

Reviewing b. "adapt the total length", I ask myself: 
How can total length be adapted, if it is the length of a _fixed_ polygon? 
Did I make a mistake by writing this comment during our last meeting?

**Comment by DrCaFr** [comment]
Unter https://openoregon.pressbooks.pub/weldsymbols/chapter/chapter-3/ habe ich folgende Diagramme gefunden:

![grafik](https://user-images.githubusercontent.com/18173895/108207747-72961100-7128-11eb-9e8c-f9a793d84237.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207770-7b86e280-7128-11eb-8133-0e2022ba68fd.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207787-817cc380-7128-11eb-9a21-0dec442019e7.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207809-86da0e00-7128-11eb-93a1-e570ae473d25.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207830-8c375880-7128-11eb-8cd8-bd6152c95a3c.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207866-98bbb100-7128-11eb-9065-b836a886dc4a.png)

![grafik](https://user-images.githubusercontent.com/18173895/108207898-a1ac8280-7128-11eb-90a2-5569c1608764.png)

**Comment by DrCaFr** [comment]
Closed by mistake. Now, reopened.

**Comment by DrCaFr** [comment]
For tomorrow's WebCo, please see draft section 8.1.2.1 in https://github.com/economidis-nick/createXSDforxMCF/raw/v3_1_1_editorial_changes/V3.1.1/Documentation_xMCF_File_v3.1.1.docx!

**Comment by economidis-nick** [comment]
> At https://openoregon.pressbooks.pub/weldsymbols/chapter/chapter-3/ , I found following diagrams:

Note that the diagrams 1, 2, 3 and 4 can already be represented using the `margin + length + spacing` definition,
as **`pitch = length + spacing`**.
The last 2 diagrams need discussion though

**Comment by DrCaFr** [comment]
Meeting of 2021-02-18 agreed to present following question to next AK 25 meeting:
> Is `<segment_list/>` (Example C) really needed, or is it sufficient to represent this by separate weld lines?

**Comment by DrCaFr** [comment]
Next meeting: Wed., 2021-03-03, 9:30-10:30 CET.

**Comment by DrCaFr** [comment]
Are attributes “begin_margin” and “end_margin” _optional_, both _defaulting to 0.0_?
(Non-optional = mandatory means that consuming systems shall emit a warning, if such an attribute is missing!)

Jürgen will provide some slide to foster discussion at next meeting.

**Comment by mweinert611** [comment]
In order to help the discussion in the next meeting, I tried to describe the 2 options or use cases in design / modelling process steps. It hopefully explains that there should not be a uncontrolled risk to compatibility of the weld definition in any of the 2 options. The (rather theoretical) incompatibilities I could imagine are listed as well...

[Intermittent_Welds_Discussion_210224.pdf](https://github.com/economidis-nick/createXSDforxMCF/files/6035165/Intermittent_Welds_Discussion_210224.pdf)

**Comment by DrCaFr** [comment]
Issue was closed by mistake.

**Comment by economidis-nick** [comment]
> Please see draft section 8.1.2.1 in https://github.com/economidis-nick/createXSDforxMCF/raw/v3_1_1_editorial_changes/V3.1.1/Documentation_xMCF_File_v3.1.1.docx!

@DrCaFr , I suggest paragraph 8.1.2.1 (Intermittent Connection Lines) should be moved explicitly under 8.2 Seamwelds and it should be renamed as "Intermittent Seamwelds" (or so). 
I don't think the concept applies to or facilitates _all_ other connection line types. So far

**Comment by DrCaFr** [comment]
> I don't think the concept applies to or facilitates _all_ other connection line types. So far

Hi @economidis-nick: Well, I _can_ imagine to have intermittent _glue_ lines, too. 
Suggest to ask the users &hellip;

**Comment by BrunsJ** [comment]
General question about the role of xMCF:
Must xMCF „only“ be able to store the variations of the intermittent welds created by a preprocessor?
In this case (begin_)margin, spacing and length are sufficient from my prospective.
Or must xMCF provide some kind of „intelligence“ to fulfill preprocessor tasks?

**Comment by BrunsJ** [comment]
I created some slides how a intermittent weld can be defined, but if xMCF is "only" ment to store welding informations the current proposal ist too complicated... We can discuss this in our meeting.

![image](https://user-images.githubusercontent.com/78599820/109765610-64022c00-7bf5-11eb-9e10-782e36879b6a.png)
![image](https://user-images.githubusercontent.com/78599820/109765683-7e3c0a00-7bf5-11eb-9d41-35883cfb6f79.png)

**Comment by DrCaFr** [comment]
Results of 2021-03-03 WebCo:
- We should formulate some requirements about minimum precision/accuracy of `<loc_list/>`. 
- length- & spacing-based definition appears to have _more_ relevance in practice than density- & num_segments-based one. (t.b.v)
- Discussion to be continued at next meeting: 18.03. 9:30-10:30 CET.

**Comment by DrCaFr** [comment]
> > I don't think the concept applies to or facilitates _all_ other connection line types. So far
> 
> Hi @economidis-nick: Well, I _can_ imagine to have intermittent _glue_ lines, too.
> Suggest to ask the users …

Decision was, to add statement "Intermittent connection lines are only applicable to weld lines, currently." to section "8.1.2.1 Intermittent Connection Lines", but not to move this section to another place in the document.

**Comment by economidis-nick** [comment]
The proposal with `begin_margin` and `end_margin` does not cover all cases of AWS found in [this comment](https://github.com/economidis-nick/createXSDforxMCF/issues/53#issuecomment-780540964).
In these cases, the first (and last) weld may have a special length by design.
The last weld may have been half-completed.

I propose this alternative definition, which caters of **all cases** described in the attached examples above.
## Alternative Definition

~~~~
|| attributes    || type     ||   use    || constraint                         ||
---------------------------------------------------------------------------------
| length         | float     | Required  |    > 0                               |
| spacing        | float     | Required  |    > 0                               |
| first_spacing  | float     | Optional  |  default = 0                         |
| first_length   | float     | Optional  |  cannot coexist with "first_spacing" |
| last_spacing   | float     | Optional  |  default = 0                         |
| last_length    | float     | Optional  |  cannot coexist with "last_spacing"  |
| keep           | Selection | Optional  |  spacing / length / density(=dflt)   |
~~~~

~~~~xml
<seamweld>
    <corner_weld base="1" technology="arc">
        <weld_position x='0.0' y='10.0' z='10.0' reference = 'no' angle='90' filler='yes' shape='convex'>
                <regular_step first_spacing="1.0" last_spacing="3.0" length="10.0" spacing="5.0" keep="spacing"/>
        </weld_position>
    </corner_weld>
</seamweld>
~~~~

![image](https://user-images.githubusercontent.com/33229181/111524678-68862300-8765-11eb-954f-5ddc1943b9ca.png)

**Comment by DrCaFr** [comment]
Results of 2021-03-18 meeting (today w/o Karin):  
- General rule: The creating system is responsible for consistent definition of the segments, not the &chi;MCF format.
- If first or last segment length shall deviate from other segment lengths, an irregular weld has to be used. 
- Excess of segments at the end of a seam weld is not allowed. 
- "keep" attribute is to be used for compensation of rounding errors. (defaults to density)
- A new attribute "max_percentage_of_compensation" is introduced:
  - If keep=length, then a warning has to be issued if adjusted spacing deviates from prescribed spacing by more than max_percentage_of_compensation.
  - Vice versa: If keep=spacing, a warning is needed if max_percentage_of_compensation is exceeded for length. 
- A new attribute "max_absolute_compensation" is introduced: Same semantics as "max_percentage_of_compensation", but deviation measured in length units, not in percents. 
- The two new attributes are mutually exclusive. Both are optional. If none is given, no check (and no warning) takes place.
- Only length and spacing are subject to compensation. Margins have always to be met exactly.

**Comment by DrCaFr** [comment]
Summary and next steps wrt. "Intermittent Welds":
- Update spec. document (Nick & Carsten)
- Update / cross check example files (3 of them) (Nick & Carsten)
- Prepare a slide for meeting in May (Matthias)

**Comment by DrCaFr** [comment]
Next meeting: April 15th, 10:00-11:00 CEST.

**Comment by DrCaFr** [comment]
Next meeting: May 4th, 9:30-10:30 CEST.

Complete AK 25 meeting is scheduled for May, 20th. Time to be defined.

**Comment by DrCaFr** [comment]
Home work: 
- consider changed strategy, i. e. _n_ is transported in &chi;MCF. 
- Think about "accuracy section", issue #62.

**Comment by DrCaFr** [comment]
Example with attribute "max_percentage_of_compensation" to be added.

**Comment by DrCaFr** [comment]
Decision: Work on this issue is finished with respect to the content. 
To-Do: Polish spec document from the change markups.

**Comment by economidis-nick** [comment]
> Example with attribute "max_percentage_of_compensation" to be added.

added in this commit [f48b7f7
](https://github.com/economidis-nick/createXSDforxMCF/commit/f48b7f788fc957f550787a2e20e789ab610c9620)

**Comment by DrCaFr** [comment]
I would suggest accepting all respective changes to the Word file and closing this issue.

However, there is a footnote on p. 110
>Thus, the reliable definition of intermittent connection lines requires a certain accuracy of this polygon<sup>21</sup>. 

where
> <sup>21</sup> Reference to corresponding new section to be added, as soon as this section comes into life. 
The new section could then be referenced e.g. by sections 8.1.2 Location, or 8.2.4.3.2 Welding Position. Corresponding issue is `https://github.com/economidis-nick/createXSDforxMCF/issues/62`. 

We could mark the footnote as "hidden text" and re-enable it, as soon as we start working on issue #62.

**Comment by DrCaFr** [comment]
Suggestion confirmed. @DrCaFr will care for accepting the pending changes to the spec and then close the issue.

**Comment by DrCaFr** [comment]
Issue closed after commit 5b1b59ee2dd84e500b29c833a80544bf36177b33.
